stdlib/public/core/IntegerParsing.swift

```diff
@@ -146,31 +45,54 @@ extension FixedWidthInteger {
   ///     `radix`.
   ///   - radix: The radix, or base, to use for converting `text` to an integer
   ///     value. `radix` must be in the range `2...36`. The default is 10.
-  @inlinable // @specializable
-  @_semantics("optimize.sil.specialize.generic.partial.never")
+  @inlinable @inline(__always)
   public init?<S: StringProtocol>(_ text: S, radix: Int = 10) {
-    _precondition(2...36 ~= radix, "Radix not in range 2...36")
-
-    if let str = text as? String, str._guts.isFastUTF8 {
-      guard let ret = str._guts.withFastUTF8 ({ utf8 -> Self? in
-        var iter = utf8.makeIterator()
-        return _parseASCII(codeUnits: &iter, radix: Self(radix))
-      }) else {
-        return nil
+    _precondition(2...36 ~= radix)
+    // Check we can load the UTF-8 in UInt64 chunks.
+    var wholeGuts = text._wholeGuts
+    // TODO: Merge?
+    let canChunkLoad: Bool
+    if S.self == String.self {
+      canChunkLoad = wholeGuts.isSmall || wholeGuts.count >= 9 // +1 for sign.
+    } else {
+      canChunkLoad = (wholeGuts.isSmall ? 16 : wholeGuts.count)
+        &- text.startIndex._encodedOffset >= 9
+    }
+    // Wrapper to handle the fact that withUF8 is defined on both
+    // String and Substring, but not on StringProtocol.
+    let withUTF8: ((UnsafeBufferPointer<UInt8>) -> Self?) -> Self? = { f in
+      if S.self == String.self {
+        var text = text as! String
+        return text.withUTF8(f)
+      } else { // StringProtocol requires no additional types conform to it.
```

Comment:
This may not be correct. I think we may technically be leaving the door open to having additional types conform to StringProtocol. In any case, it is worth considering if we want to expose this in an `inlinable` function.